panic: runtime error: index out of range

Hi! I've encountered a strange error in backup recovery tool WAL-G using Zstd compression.
Here's a call stack. Who do you think, can this be caused by corrupted data on WAL-G side or is it a problem in Zstd bindings?
Here https://github.com/DataDog/zstd/blob/1.x/zstd_stream.go#L240 code does not check that ```src``` has at least one byte.

### What version of Go are you using (`go version`)?
1.10


### What operating system and processor architecture are you using (`go env`)?
Linux 86_64


### What did you do?

wal-g backup-fetch of 11Tb database

### What did you expect to see?


### What did you see instead?

```
panic: runtime error: index out of range

goroutine 4949956 [running]:
github.com/x4m/wal-g/vendor/github.com/DataDog/zstd.(*reader).Read(0xc460b9c360, 0xc43aa28000, 0x1400000, 0x1400000, 0x1400000, 0x0, 0x0)
        /home/robot-gerrit/workspace/wal-g-build-deb/src/github.com/x4m/wal-g/obj-x86_64-linux-gnu/src/github.com/x4m/wal-g/vendor/github.com/DataDog/zstd/zstd_stream.go:240 +0x924
github.com/x4m/wal-g.fastCopy(0xf123a0, 0xc44de66c80, 0x7f5e745d9198, 0xc460b9c360, 0xc460b9c360, 0xc4464e5e60, 0xc420c3bf28)
        /home/robot-gerrit/workspace/wal-g-build-deb/src/github.com/x4m/wal-g/obj-x86_64-linux-gnu/src/github.com/x4m/wal-g/
```

--- a/zstd_stream.go
+++ b/zstd_stream.go
@@ -14,6 +14,8 @@ import (
 	"unsafe"
 )
 
+var errShortRead = errors.New("short read")
+
 // Writer is an io.WriteCloser that zstd-compresses its input.
 type Writer struct {
 	CompressionLevel int
@@ -224,7 +226,7 @@ func (r *reader) Read(p []byte) (int, error) {
 		src := r.compressionBuffer
 		reader := r.underlyingReader
 		n, err := TryReadFull(reader, src[r.compressionLeft:])
-		if err != nil && err != ErrShortRead { // Handle underlying reader errors first
+		if err != nil && err != errShortRead { // Handle underlying reader errors first
 			return 0, fmt.Errorf("failed to read from underlying reader: %s", err)
 		} else if n == 0 && r.compressionLeft == 0 {
 			return got, io.EOF
@@ -271,7 +273,7 @@ func (r *reader) Read(p []byte) (int, error) {
 
 // TryReadFull reads buffer just as ReadFull does
 // Here we expect that buffer may end and we do not return ErrUnexpectedEOF as ReadAtLeast does.
-// We return ErrShortRead instead to distinguish short reads and failures.
+// We return errShortRead instead to distinguish short reads and failures.
 // We cannot use ReadFull/ReadAtLeast because it masks Reader errors, such as network failures
 // and causes panic instead of error.
 func TryReadFull(r io.Reader, buf []byte) (n int, err error) {
@@ -283,9 +285,7 @@ func TryReadFull(r io.Reader, buf []byte) (n int, err error) {
 	if n == len(buf) && err == io.EOF {
 		err = nil // EOF at the end is somewhat expected
 	} else if err == io.EOF {
-		err = ErrShortRead
+		err = errShortRead
 	}
 	return
 }
-
-var ErrShortRead = errors.New("short read")

--- a/zstd_stream.go
+++ b/zstd_stream.go
@@ -8,6 +8,7 @@ package zstd
 */
 import "C"
 import (
+	"errors"
 	"fmt"
 	"io"
 	"unsafe"
@@ -222,11 +223,11 @@ func (r *reader) Read(p []byte) (int, error) {
 		// Populate src
 		src := r.compressionBuffer
 		reader := r.underlyingReader
-		n, err := io.ReadFull(reader, src[r.compressionLeft:])
-		if err == io.EOF && r.compressionLeft == 0 {
-			return got, io.EOF
-		} else if err != nil && err != io.EOF && err != io.ErrUnexpectedEOF {
+		n, err := TryReadFull(reader, src[r.compressionLeft:])
+		if err != nil && err != ErrShortRead { // Handle underlying reader errors first
 			return 0, fmt.Errorf("failed to read from underlying reader: %s", err)
+		} else if n == 0 && r.compressionLeft == 0 {
+			return got, io.EOF
 		}
 		src = src[:r.compressionLeft+n]
 
@@ -267,3 +268,24 @@ func (r *reader) Read(p []byte) (int, error) {
 	}
 	return got, nil
 }
+
+// TryReadFull reads buffer just as ReadFull does
+// Here we expect that buffer may end and we do not return ErrUnexpectedEOF as ReadAtLeast does.
+// We return ErrShortRead instead to distinguish short reads and failures.
+// We cannot use ReadFull/ReadAtLeast because it masks Reader errors, such as network failures
+// and causes panic instead of error.
+func TryReadFull(r io.Reader, buf []byte) (n int, err error) {
+	for n < len(buf) && err == nil {
+		var nn int
+		nn, err = r.Read(buf[n:])
+		n += nn
+	}
+	if n == len(buf) && err == io.EOF {
+		err = nil // EOF at the end is somewhat expected
+	} else if err == io.EOF {
+		err = ErrShortRead
+	}
+	return
+}
+
+var ErrShortRead = errors.New("short read")

--- a/zstd_stream_test.go
+++ b/zstd_stream_test.go
@@ -171,3 +171,18 @@ func BenchmarkStreamDecompression(b *testing.B) {
 		}
 	}
 }
+
+type breakingReader struct {
+}
+
+func (r *breakingReader) Read(p []byte) (int, error) {
+	return len(p) - 1, io.ErrUnexpectedEOF
+}
+
+func TestUnexpectedEOFHandling(t *testing.T) {
+	r := NewReader(&breakingReader{})
+	_, err := r.Read(make([]byte, 1024))
+	if err == nil {
+		t.Error("Underlying error was handled silently")
+	}
+}

